// typecomp/dist/workspace.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const TS = require("typescript");
const project_1 = require("./project");
const helpers_1 = require("./helpers");
class Workspace {
    constructor() {
        this.projects = new Map();
        this.registry = TS.createDocumentRegistry(false, process.cwd());
    }
    projectFor(filename) {
        const root = helpers_1.getRoot(filename);
        if (root === null) {
            throw new Error(`${filename} has no associated TypeScript configuration`);
        }
        let project = this.projects.get(root);
        if (project === undefined) {
            project = new project_1.Project(root, this.registry);
            this.projects.set(root, project);
        }
        return project;
    }
    index(filename) {
        this.projectFor(filename).index(filename);
    }
    diagnose(filename) {
        return this.projectFor(filename).diagnose(filename);
    }
    compile(filename) {
        return this.projectFor(filename).compile(filename);
    }
}
exports.Workspace = Workspace;
